teams/vsa/engineering/forms-system/custom_component_multiple.md

```diff
@@ -0,0 +1,138 @@
+# Extending RJSF Behavior Using a Custom React Component
+
+(Based mostly on [bradl3yC](https://github.com/bradl3yC)'s write-up of his solution)
+
+## The Problem:
+Using a custom component requires that you pass in a schema for that component.  It essentially maps a small slice of form state to the component.  Because the same component is used twice but we are passing in a different schema, each of those components, each of those only has access to its own individual schema/form state. (Read [Background: The Original Problem](#background-the-original-problem) for further details.)
+
+## The Solution/Hack:
```

Comment:
I feel a little strange about the most comprehensive document to be highlighting solutions as hacks. It feels like we're promoting an attitude of "we're stuck with this technology, so do what you gotta do" when we definitely have the ability to influence and directly contribute to the forms system.

I don't know, just thinking how/if we could we define guidance on how to identify when you need a custom component, what the right way to build it is vs. the temporary/urgent way (I'm thinking/hoping that's what we mean here by "back" because it implies accumulating tech debt IMO), and how to identify when _not_ to use the forms system, how to provide feedback to designers about steering design towards ease of implementation ("we can't do this out-of-the-box with the forms system so we'd have to build a new widget with a LOE of X. However, we could change the design to use a X, in which case this would be easy"). Maybe if we could define those things first we could segue into this approach. 

Also, a table of contents at this root of this directory could help us navigate these docs and their objectives.

Reply:
I've been talking about exactly to my designers - the idea that it's best to build within the form system and if a design can't accommodate that, then to identify it in advance and to go through VSP to recommend changing the form system instead of hacking outside of the form system. In order to be able to do this, both designers and developers should be able to identify whether or not something will be harder / if we can't do it "out of the box." My designers sometimes struggle with this because they see examples in other forms and aren't sure if small tweaks will add hugely to the LOE / make something custom. So we need developers that can really dig into early designs and let designers know when something looks problematic. Or some process for this. I've suggested PMs consciously writing tickets where devs look at early designs and flag anything that might be problematic.

Reply:
@ncksllvn and @sshein The scope of this particular item is limited to documenting patterns that we have used. That is a first step (data collection for empirical analysis), not the end goal. The analysis of the patterns is the next step (read-through with VSP), then after that we would get into design and solutions. What you are seeing here is essentially a capture of the AS-IS state so that empirical analysis can be conducted at the technical level between developers in VSA, VSP, and possibly VAOS:

I believe what both of you are looking for might be covered in my pending check-in which has some proposed solutions in a **wish list section**, especially the first two items:

- A visual design-oriented catalog that shows the forms components integrated on test pages (100% traceability).

- A live playground feature similar to the [one for the original RSFJ](https://rjsf-team.github.io/react-jsonschema-form/).

- Documentation on *formConfig* options such as *hideIf, hideOnReview, hideOnReviewIfFalse, updateFormData, updateSchema, expandUnder* (including limitations we have run into with HideIf, expandIf, and why.)

- Visible backlog of VAFS/RJSF items in VSP that VSA developers can see so that we have transparency into what is known and not known as far as defect, features, workarounds, etc. WHEN or IF we get any of these improvements is a totally different story, but we at least need to keep track of the issues so that we don't keep stepping into them. When we get into the how/solutioning, VSA developers would be expected to work with VSP (joint effort, not something that would be expected to be handed down).

- More data mined from Slack conversations and the existing forms in source control. For example, forms 2346, 526EZ, 686, 686c-674, 996 are listed under [src/applications/disability-benefits](https://github.com/department-of-veterans-affairs/vets-website/tree/master/src/applications/disability-benefits). Would like to have an empirical analysis that summarizes *all the forms* we have built in va.<span/>gov so far and identify what patterns we are using. Are 80% just out-of-the box, and 20% special cases? Would like to know which ones are special, and for those special cases, what implementation patterns were used.

Reply:
We should also come up with some kind of rating system on patterns we have found, such as "recommended", "workaround", "obsolete", "avoid"

Reply:
> Also, a table of contents at this root of this directory could help us navigate these docs and their objectives.

@ncksllvn the root readme.md does this; however, I'll make it more obvious (put things under Contents, rename Other Guidance to External Links, etc.)

Reply:
100% visibility across design and development so everyone knows what's going on will solve a lot of confusion.